src/centralstore/incremental/NodeManager.cpp

```diff
@@ -0,0 +1,254 @@
+    RelationBlock *newRelation = NULL;
+    if (source.edgeRef == 0 || destination.edgeRef == 0 ||
+        !source.searchRelation(destination)) {  // certainly a new relation block needed
+        newRelation = RelationBlock::add(source, destination);
+        if (newRelation) {
+            source.updateRelation(newRelation);
+            destination.updateRelation(newRelation);
+        } else {
+            node_manager_logger.error("Something went wrong while adding the new edge/relation source = " +
+                                      std::string(source.id) + " destination = " + std::string(destination.id));
+        }
+
+    } else {
+        node_manager_logger.warn("Relation/Edge already exist for source = " + std::string(source.id) +
+                                 " destination = " + std::string(destination.id));
+    }
+    return 0;
+}
+
+NodeBlock *NodeManager::addNode(std::string nodeId) {
+    unsigned int assignedNodeIndex;
+    if (this->nodeIndex.find(nodeId) == this->nodeIndex.end()) {
+        node_manager_logger.debug("NodeId not found in index for node ID " + nodeId);
+        NodeBlock *sourceBlk = new NodeBlock(nodeId, this->nextNodeIndex * NodeBlock::BLOCK_SIZE);
+        this->nodeIndex.insert({nodeId, this->nextNodeIndex});
+        assignedNodeIndex = this->nextNodeIndex;
+        this->nextNodeIndex++;
+        sourceBlk->save();
+        return sourceBlk;
+    } else {
+        node_manager_logger.debug("NodeId found in index for node ID " + nodeId);
+        return this->get(nodeId);
+    }
+}
+
+std::pair<NodeBlock, NodeBlock> NodeManager::addEdge(std::pair<std::string, std::string> edge) {
+    NodeBlock *sourceNode = this->addNode(edge.first);
+    NodeBlock *destNode = this->addNode(edge.second);
+    unsigned int relationAddr = this->addRelation(*sourceNode, *destNode);
+
+    node_manager_logger.debug("DEBUG: Source DB block address " + std::to_string(sourceNode->addr) +
+                              " Destination DB block address " + std::to_string(destNode->addr));
+    std::pair<NodeBlock, NodeBlock> returnVal = {*sourceNode, *destNode};
+    delete sourceNode; // to prevent memory leaks
+    delete destNode;
+    return returnVal;
+}
+
+int NodeManager::dbSize(std::string path) {
+    /*
+        The structure stat contains at least the following members:
+        st_dev     ID of device containing file
+        st_ino     file serial number
+        st_mode    mode of file (see below)
+        st_nlink   number of links to the file
+        st_uid     user ID of file
+        st_gid     group ID of file
+        st_rdev    device ID (if file is character or block special)
+        st_size    file size in bytes (if file is a regular file)
+        st_atime   time of last access
+        st_mtime   time of last data modification
+        st_ctime   time of last status change
+        st_blksize a filesystem-specific preferred I/O block size for
```

Comment:
In this case we should have some example block sizes defined as constants/enumeration.

Reply:
This comment is actually an extract from [this wiki](https://en.wikipedia.org/wiki/Stat_(system_call)), As a reference for what's in the `stat` structure.
Shall I remove the comment ?

Reply:
We can keep the comment. In addition can you please add the link URL to the code as a comment so that we can track back when needed?

Reply:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/771d0ebe5f06b0a8fa6801ae04b86f8aa0bc2241

Reply:
Done